File: pdu/python/drone_srv_msgs/pdu_conv_DroneGetStateResponse.py

import struct
from .pdu_pytype_DroneGetStateResponse import DroneGetStateResponse
from ..pdu_utils import *
from .. import binary_io

# dependencies for the generated Python class
from ..geometry_msgs.pdu_conv_Pose import *
from ..hako_msgs.pdu_conv_HakoBatteryStatus import *



def pdu_to_py_DroneGetStateResponse(binary_data: bytearray) -> DroneGetStateResponse:
    py_obj = DroneGetStateResponse()
    meta_parser = binary_io.PduMetaDataParser()
    meta = meta_parser.load_pdu_meta(binary_data)
    if meta is None:
        raise ValueError("Invalid PDU binary data: MetaData not found or corrupted")
    binary_read_recursive_DroneGetStateResponse(meta, binary_data, py_obj, binary_io.PduMetaData.PDU_META_DATA_SIZE)
    return py_obj


def binary_read_recursive_DroneGetStateResponse(meta: binary_io.PduMetaData, binary_data: bytearray, py_obj: DroneGetStateResponse, base_off: int):
    # array_type: single 
    # data_type: primitive 
    # member_name: ok 
    # type_name: bool 
    # offset: 0 size: 4 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 0, 4)
    py_obj.ok = binary_io.binTovalue("bool", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: is_ready 
    # type_name: bool 
    # offset: 4 size: 4 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 4, 4)
    py_obj.is_ready = binary_io.binTovalue("bool", bin)
    
    # array_type: single 
    # data_type: struct 
    # member_name: current_pose 
    # type_name: geometry_msgs/Pose 
    # offset: 8 size: 56 
    # array_len: 1

    tmp_py_obj = Pose()
    binary_read_recursive_Pose(meta, binary_data, tmp_py_obj, base_off + 8)
    py_obj.current_pose = tmp_py_obj
    
    # array_type: single 
    # data_type: struct 
    # member_name: battery_status 
    # type_name: hako_msgs/HakoBatteryStatus 
    # offset: 64 size: 32 
    # array_len: 1

    tmp_py_obj = HakoBatteryStatus()
    binary_read_recursive_HakoBatteryStatus(meta, binary_data, tmp_py_obj, base_off + 64)
    py_obj.battery_status = tmp_py_obj
    
    # array_type: single 
    # data_type: primitive 
    # member_name: mode 
    # type_name: string 
    # offset: 96 size: 128 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 96, 128)
    py_obj.mode = binary_io.binTovalue("string", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: message 
    # type_name: string 
    # offset: 224 size: 128 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 224, 128)
    py_obj.message = binary_io.binTovalue("string", bin)
    
    return py_obj


def py_to_pdu_DroneGetStateResponse(py_obj: DroneGetStateResponse) -> bytearray:
    binary_data = bytearray()
    base_allocator = DynamicAllocator(False)
    bw_container = BinaryWriterContainer(binary_io.PduMetaData())
    binary_write_recursive_DroneGetStateResponse(0, bw_container, base_allocator, py_obj)

    # メタデータの設定
    total_size = base_allocator.size() + bw_container.heap_allocator.size() + binary_io.PduMetaData.PDU_META_DATA_SIZE
    bw_container.meta.total_size = total_size
    bw_container.meta.heap_off = binary_io.PduMetaData.PDU_META_DATA_SIZE + base_allocator.size()

    # binary_data のサイズを total_size に調整
    if len(binary_data) < total_size:
        binary_data.extend(bytearray(total_size - len(binary_data)))
    elif len(binary_data) > total_size:
        del binary_data[total_size:]

    # メタデータをバッファにコピー
    binary_io.writeBinary(binary_data, 0, bw_container.meta.to_bytes())

    # 基本データをバッファにコピー
    binary_io.writeBinary(binary_data, bw_container.meta.base_off, base_allocator.to_array())

    # ヒープデータをバッファにコピー
    binary_io.writeBinary(binary_data, bw_container.meta.heap_off, bw_container.heap_allocator.to_array())

    return binary_data

def binary_write_recursive_DroneGetStateResponse(parent_off: int, bw_container: BinaryWriterContainer, allocator, py_obj: DroneGetStateResponse):
    # array_type: single 
    # data_type: primitive 
    # member_name: ok 
    # type_name: bool 
    # offset: 0 size: 4 
    # array_len: 1
    type = "bool"
    off = 0

    
    bin = binary_io.typeTobin(type, py_obj.ok)
    bin = get_binary(type, bin, 4)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: is_ready 
    # type_name: bool 
    # offset: 4 size: 4 
    # array_len: 1
    type = "bool"
    off = 4

    
    bin = binary_io.typeTobin(type, py_obj.is_ready)
    bin = get_binary(type, bin, 4)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: struct 
    # member_name: current_pose 
    # type_name: geometry_msgs/Pose 
    # offset: 8 size: 56 
    # array_len: 1
    type = "Pose"
    off = 8

    binary_write_recursive_Pose(parent_off + off, bw_container, allocator, py_obj.current_pose)
    
    # array_type: single 
    # data_type: struct 
    # member_name: battery_status 
    # type_name: hako_msgs/HakoBatteryStatus 
    # offset: 64 size: 32 
    # array_len: 1
    type = "HakoBatteryStatus"
    off = 64

    binary_write_recursive_HakoBatteryStatus(parent_off + off, bw_container, allocator, py_obj.battery_status)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: mode 
    # type_name: string 
    # offset: 96 size: 128 
    # array_len: 1
    type = "string"
    off = 96

    
    bin = binary_io.typeTobin(type, py_obj.mode)
    bin = get_binary(type, bin, 128)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: message 
    # type_name: string 
    # offset: 224 size: 128 
    # array_len: 1
    type = "string"
    off = 224

    
    bin = binary_io.typeTobin(type, py_obj.message)
    bin = get_binary(type, bin, 128)
    allocator.add(bin, expected_offset=parent_off + off)
    

if __name__ == "__main__":
    import sys
    import json

    def print_usage():
        print(f"Usage: python -m pdu.python.pdu_conv_DroneGetStateResponse <read|write> [args...]")
        print(f"  read <input_binary_file> <output_json_file>")
        print(f"  write <input_json_file> <output_binary_file>")

    if len(sys.argv) < 2:
        print_usage()
        sys.exit(1)

    command = sys.argv[1]

    if command == "read":
        if len(sys.argv) != 4:
            print_usage()
            sys.exit(1)
        
        binary_filepath = sys.argv[2]
        output_json_filepath = sys.argv[3]

        with open(binary_filepath, "rb") as f:
            binary_data = bytearray(f.read())
        
        py_obj = pdu_to_py_DroneGetStateResponse(binary_data)
        
        with open(output_json_filepath, "w") as f:
            f.write(py_obj.to_json())

    elif command == "write":
        if len(sys.argv) != 4:
            print_usage()
            sys.exit(1)

        input_json_filepath = sys.argv[2]
        output_binary_filepath = sys.argv[3]

        with open(input_json_filepath, "r") as f:
            json_str = f.read()
        
        py_obj = DroneGetStateResponse.from_json(json_str)
        
        binary_data = py_to_pdu_DroneGetStateResponse(py_obj)

        with open(output_binary_filepath, "wb") as f:
            f.write(binary_data)

    else:
        print(f"Unknown command: {command}")
        print_usage()
        sys.exit(1)

File: pdu/javascript/drone_srv_msgs/pdu_conv_DroneGetStateResponse.js
import * as PduUtils from '../pdu_utils.js';
import { DroneGetStateResponse } from './pdu_jstype_DroneGetStateResponse.js';
import { Pose } from '../geometry_msgs/pdu_jstype_Pose.js';
import { binary_read_recursive_Pose, binary_write_recursive_Pose } from '../geometry_msgs/pdu_conv_Pose.js';
import { HakoBatteryStatus } from '../hako_msgs/pdu_jstype_HakoBatteryStatus.js';
import { binary_read_recursive_HakoBatteryStatus, binary_write_recursive_HakoBatteryStatus } from '../hako_msgs/pdu_conv_HakoBatteryStatus.js';


/**
 * Deserializes a binary PDU into a DroneGetStateResponse object.
 * @param {ArrayBuffer} binary_data
 * @returns { DroneGetStateResponse }
 */
export function pduToJs_DroneGetStateResponse(binary_data) {
    const js_obj = new DroneGetStateResponse();
    const meta_parser = new PduUtils.PduMetaDataParser();
    const meta = meta_parser.load_pdu_meta(binary_data);
    if (meta === null) {
        throw new Error("Invalid PDU binary data: MetaData not found or corrupted");
    }
    binary_read_recursive_DroneGetStateResponse(meta, binary_data, js_obj, meta.base_off);
    return js_obj;
}

export function binary_read_recursive_DroneGetStateResponse(meta, binary_data, js_obj, base_off) {
    const view = new DataView(binary_data);
    const littleEndian = true;
    // member: ok, type: bool (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 0, 4);
        js_obj.ok = PduUtils.binToValue("bool", bin);
    }
    
    // member: is_ready, type: bool (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 4, 4);
        js_obj.is_ready = PduUtils.binToValue("bool", bin);
    }
    
    // member: current_pose, type: geometry_msgs/Pose (struct)

    {
        const tmp_obj = new Pose();
        binary_read_recursive_Pose(meta, binary_data, tmp_obj, base_off + 8);
        js_obj.current_pose = tmp_obj;
    }
    
    // member: battery_status, type: hako_msgs/HakoBatteryStatus (struct)

    {
        const tmp_obj = new HakoBatteryStatus();
        binary_read_recursive_HakoBatteryStatus(meta, binary_data, tmp_obj, base_off + 64);
        js_obj.battery_status = tmp_obj;
    }
    
    // member: mode, type: string (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 96, 128);
        js_obj.mode = PduUtils.binToValue("string", bin);
    }
    
    // member: message, type: string (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 224, 128);
        js_obj.message = PduUtils.binToValue("string", bin);
    }
    
    return js_obj;
}

/**
 * Serializes a DroneGetStateResponse object into a binary PDU.
 * @param { DroneGetStateResponse } js_obj
 * @returns {ArrayBuffer}
 */
export function jsToPdu_DroneGetStateResponse(js_obj) {
    const base_allocator = new PduUtils.DynamicAllocator();
    const bw_container = new PduUtils.BinaryWriterContainer(new PduUtils.PduMetaData());

    binary_write_recursive_DroneGetStateResponse(0, bw_container, base_allocator, js_obj);

    const base_data_size = base_allocator.size();
    const heap_data_size = bw_container.heap_allocator.size();
    
    bw_container.meta.heap_off = PduUtils.PDU_META_DATA_SIZE + base_data_size;
    bw_container.meta.total_size = bw_container.meta.heap_off + heap_data_size;

    const final_buffer = new ArrayBuffer(bw_container.meta.total_size);
    const final_view = new Uint8Array(final_buffer);

    PduUtils.writeBinary(final_view, 0, bw_container.meta.to_bytes());
    PduUtils.writeBinary(final_view, bw_container.meta.base_off, base_allocator.toArray());
    PduUtils.writeBinary(final_view, bw_container.meta.heap_off, bw_container.heap_allocator.toArray());

    return final_buffer;
}

export function binary_write_recursive_DroneGetStateResponse(parent_off, bw_container, allocator, js_obj) {
    const littleEndian = true;
    // member: ok, type: bool (primitive)

    
    {
        const bin = PduUtils.typeToBin("bool", js_obj.ok, 4);
        allocator.add(bin, parent_off + 0);
    }
    
    // member: is_ready, type: bool (primitive)

    
    {
        const bin = PduUtils.typeToBin("bool", js_obj.is_ready, 4);
        allocator.add(bin, parent_off + 4);
    }
    
    // member: current_pose, type: geometry_msgs/Pose (struct)

    {
        binary_write_recursive_Pose(parent_off + 8, bw_container, allocator, js_obj.current_pose);
    }
    
    // member: battery_status, type: hako_msgs/HakoBatteryStatus (struct)

    {
        binary_write_recursive_HakoBatteryStatus(parent_off + 64, bw_container, allocator, js_obj.battery_status);
    }
    
    // member: mode, type: string (primitive)

    
    {
        const bin = PduUtils.typeToBin("string", js_obj.mode, 128);
        allocator.add(bin, parent_off + 96);
    }
    
    // member: message, type: string (primitive)

    
    {
        const bin = PduUtils.typeToBin("string", js_obj.message, 128);
        allocator.add(bin, parent_off + 224);
    }
    
}
